parser_tests/update_snapshots.py

```diff
@@ -0,0 +1,68 @@
+import json
+import os
+import sys
+from unittest import mock
+from urllib.request import urlopen
+
+from config import parsers
+from parser_tests.regression_test import get_canteen_url, \
+    get_snapshot_result_path, \
+    get_snapshot_website_path, \
+    parse_mocked, parsers_to_test
+
+base_directory = os.path.dirname(os.path.realpath(__file__))
+
+
+def generate_all_snapshots():
+    for parser, canteen in parsers_to_test:
+        generate_snapshot(parser, canteen)
+
+
+def generate_snapshot(parser, canteen):
+    intercepted_requests = {}
+
+    def intercept_request(requested_url):
+        response = urlopen(requested_url)
+
+        html = response.read()
+        intercepted_requests[requested_url] = html.decode('utf-8')
+
+        # cannot return original response, because read() has side effects on it
+        return urlopen(requested_url)
+
+    with mock.patch('urllib.request.urlopen', intercept_request):
+        url = get_canteen_url(parser, canteen)
+        parsers[parser].parse(url, canteen, 'full.xml')
+
+    # TODO: Improve mock so that parsers are free to use request library of their choice.
+    if len(intercepted_requests) == 0:
+        raise RuntimeError("Could not intercept the HTTP calls from parser '{}'. "
+                           "Please use `from urllib import request` "
+                           "and `request.urlopen`.".format(parser))
+
+    snapshot_website_path = get_snapshot_website_path(parser, canteen)
+    with open(snapshot_website_path, 'w', encoding='utf-8') as file:
+        json.dump(intercepted_requests, file, indent=4)
+
+    snapshot_result_path = get_snapshot_result_path(parser, canteen)
+    with open(snapshot_result_path, 'w', encoding='utf-8') as result_file:
+        result = parse_mocked(parser, canteen)
```

Comment:
What if the parser failed? Then the source file was already updated, but the result file is not.

Reply:
It will hopefully fail by raising an exception, right? If you generate a snapshot for a specific parser, that should be enough. Maybe we can add a try mechanism when generating all parsers' snapshots and inform the user about which ones failed.

And even if only the website is updated and not the result: This means something is wrong with the parser. Once that's fixed, you could rerun the snapshot update.

I agree with you that this is a tricky edge case. We will have to observe this.

Reply:
The exception is the problem, as it aborts the function and `result_file.write(result)` is never called but `son.dump(…)` was already. However, the fix is really simple, just put the `result = parse_mocked(parser, canteen)` call above the writing of the source (json) file.

Reply:
That's really elegant. :+1: